### PR TITLE
Performance 1

### PR DESCRIPTION
**Pull request by ds-sloth**
Hi Wohlstand,

Please review this PR. Features:

(1) uses much simpler and more efficient table data structure instead of the overly general loose quadtree
(2) includes a new table for BGOs
(3) allows layer tables to easily split from the main table when they start moving, and join back to the main table when they stop

Remaining tasks for the future:

(1) make tables for PEZ and warps
(2) align the tables to the standard SMBX64 32x32 grid that most levels follow
(3) improve NPC performance

I think that (1) and (2) can happen directly in the main branch, and I can open another PR for (3) in the future.

### CODE REVIEW

**Comment by Wohlstand** [comment]
Gonna get a shot to fix the Windows build first, booting my VM...

File: src/player/player_update.cpp [hunk 1]
```diff
@@ -3117,11 +3120,14 @@ void UpdatePlayer()
...
-                for(B = 1; B <= numBackground; B++)
+                for(int B : treeBackgroundQuery(Player[A].Location, SORTMODE_NONE))
                 {
+                    if(B > numBackground)
+                        continue;
+
                     if(BackgroundFence[Background[B].Type] && (!g_compatibility.fix_climb_invisible_fences || !Background[B].Hidden))
                     {
-                        if(CheckCollision(Player[A].Location, Background[B].Location))
+                        // if(CheckCollision(Player[A].Location, Background[B].Location))
```
Comment (Wohlstand) [inline src/player/player_update.cpp:3130]: Why did you remove the collision check with the vine BGOs? Are you absolutely sure player won't collide them outside the actual collision?

**Review by Wohlstand — COMMENTED** [review]
All seems fine, but I made minor edits to the code style. But, I have one small question.

File: src/player/player_update.cpp (hunk 1 above)
Comment (ds-sloth) [reply src/player/player_update.cpp:3130]: I agree, that doesn't seem right. Let me look closer...

Comment (ds-sloth) [reply src/player/player_update.cpp:3130]: Ah, right. Refer to the tempLocation collision below it, it was just a duplicated check that increased computational efficiency when collision was unlikely, but reduces computational efficiency when collision is likely.

**Review by Wohlstand — APPROVED** [review]
So, now seems fine to merge!
